### DISCUSSION TITLE
[Question] [MLE] Border router initialisation

### DISCUSSION THREAD

**Discussion by caipiblack** [openthread/openthread#9727] [Q&A] [answered]
Hello

_This issue is just for questions about MLE._

**First, the story:**

We find out that when we create a thread network (1) if just after we commission a Matter (Thread) device, it is always failing on the first mDNS attempt to discover the device. (FindOperational step), then on the second attempt (45s later) it discovers the device in less that 1 seconds. So we decide to investigate to know why the first mDNS discover whas always failing.

We identify that the problem appear only if we create a thread network (1) and immediately commission a matter device. 

(1) We create the network with these commands:
```
ot-ctl thread stop
ot-ctl dataset init new
ot-ctl dataset channel $1
ot-ctl dataset commit active
ot-ctl thread start
```

If we wait for the "network to be ready" before commissioning the matter device, it seems ok:
1. We start the thread network
2. We wait initialisation complete (we considere it as complete when `ot-ctl state` = "child","router" or "leader")
3. We get the dataset
4. We start the commissioning of a matter device

And now I have some questions about this experiment. The questions will be based on the MLE logs of otbr-agent mixed with matter-commissioniner logs (but full logs are attached separatly)

```
Dec 18 10:36:07 otbr-agent[783]: 01:22:38.478 [N] Mle-----------: Role leader -> detached
Dec 18 10:36:07 otbr-agent[783]: 01:22:38.479 [N] Mle-----------: Role detached -> disabled
Dec 18 10:36:08 otbr-agent[783]: 01:22:38.640 [N] Mle-----------: Role disabled -> detached
Dec 18 10:36:08 otbr-agent[783]: 01:22:38.642 [I] Mle-----------: Attempt to become router
Dec 18 10:36:08 otbr-agent[783]: 01:22:38.642 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:12 otbr-agent[783]: 01:22:43.297 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:18 otbr-agent[783]: 01:22:48.581 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:22 otbr-agent[783]: 01:22:53.259 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:28 otbr-agent[783]: 01:22:58.689 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:33 otbr-agent[783]: 01:23:03.756 [I] Mle-----------: Send Link Request (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:34 otbr-agent[783]: 01:23:05.556 [N] Mle-----------: RLOC16 6000 -> fffe
Dec 18 10:36:34 otbr-agent[783]: 01:23:05.562 [I] Mle-----------: AttachState Idle -> Start
Dec 18 10:36:35 otbr-agent[783]: 01:23:05.877 [N] Mle-----------: Attach attempt 1, AnyPartition reattaching with Active Dataset
Dec 18 10:36:35 otbr-agent[783]: 01:23:05.878 [I] Mle-----------: AttachState Start -> ParentReq
Dec 18 10:36:35 otbr-agent[783]: 01:23:05.878 [I] Mle-----------: Send Parent Request to routers (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:36 otbr-agent[783]: 01:23:06.628 [I] Mle-----------: Send Parent Request to routers (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:36 otbr-agent[783]: 01:23:07.380 [I] Mle-----------: Send Parent Request to routers and REEDs (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:38 otbr-agent[783]: 01:23:08.631 [I] Mle-----------: Send Parent Request to routers and REEDs (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:39 otbr-agent[783]: 01:23:09.882 [I] Mle-----------: Send Parent Request to routers and REEDs (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:40 otbr-agent[783]: 01:23:11.132 [I] Mle-----------: Send Parent Request to routers and REEDs (ff02:0:0:0:0:0:0:2)
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.383 [I] Mle-----------: AttachState ParentReq -> Idle
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.383 [N] Mle-----------: RLOC16 fffe -> 6000
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.389 [N] Mle-----------: Role detached -> leader
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.389 [N] Mle-----------: Partition ID 0xb9222bf
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.390 [I] Mle-----------: Send Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:41 otbr-agent[783]: 01:23:12.400 [I] Settings------: ... pid:0xb9222bf, mlecntr:0x5e37, maccntr:0x57e1, mliid:1a6d1c20f92217a4}
Dec 18 10:36:42 otbr-agent[783]: 01:23:13.308 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:42 otbr-agent[783]: 01:23:13.426 [I] Mle-----------: Send Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:43 otbr-agent[783]: 01:23:14.114 [I] Mle-----------: Send Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:44 otbr-agent[783]: 01:23:15.344 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:46 otbr-agent[783]: 01:23:17.476 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:55 otbr-agent[783]: 01:23:26.330 [I] Mle-----------: Send Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:36:56 otbr-agent[783]: 01:23:26.966 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:37:08 otbr-agent[783]: 01:23:38.711 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:37:38 otbr-agent[783]: 01:24:08.959 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:38:10 otbr-agent[783]: 01:24:41.329 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:38:38 otbr-agent[783]: 01:25:08.774 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:39:09 otbr-agent[783]: 01:25:39.788 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:39:39 otbr-agent[783]: 01:26:10.275 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:40:09 otbr-agent[783]: 01:26:39.963 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)

12-18 10:40:33.814  7811  8060 D CTL     : Performing next commissioning step 'ReadCommissioningInfo'
12-18 10:40:34.219  7811  8060 D CTL     : Performing next commissioning step 'ArmFailSafe'
12-18 10:40:34.385  7811  8060 D CTL     : Performing next commissioning step 'ConfigRegulatory'
12-18 10:40:34.570  7811  8060 D CTL     : Performing next commissioning step 'SendPAICertificateRequest'
12-18 10:40:34.962  7811  8060 D CTL     : Performing next commissioning step 'SendDACCertificateRequest'
12-18 10:40:35.334  7811  8060 D CTL     : Performing next commissioning step 'SendAttestationRequest'
12-18 10:40:35.635  7811  8060 D CTL     : Performing next commissioning step 'AttestationVerification'
12-18 10:40:38.259  7811  7811 D CTL     : Performing next commissioning step 'SendOpCertSigningRequest'
12-18 10:40:38.806  7811  8060 D CTL     : Performing next commissioning step 'ValidateCSR'
12-18 10:40:38.854  7811  8060 D CTL     : Performing next commissioning step 'GenerateNOCChain'
12-18 10:40:38.883  7811  8060 D CTL     : Performing next commissioning step 'SendTrustedRootCert'
12-18 10:40:39.247  7811  8060 D CTL     : Performing next commissioning step 'SendNOC'
12-18 10:40:39.668  7811  8060 D CTL     : Performing next commissioning step 'ThreadNetworkSetup'
12-18 10:40:39.856  7811  8060 D CTL     : Performing next commissioning step 'FailsafeBeforeThreadEnable'
12-18 10:40:40.042  7811  8060 D CTL     : Performing next commissioning step 'ThreadNetworkEnable'
12-18 10:40:46.899  7811  8094 D CTL     : Performing next commissioning step 'FindOperational'


Dec 18 10:40:52 otbr-agent[783]: 01:27:22.697 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.687 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.687 [N] Mle-----------: Different partition (peer:1499419400, local:194126527)
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.687 [I] Mle-----------: AttachState Idle -> Start
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.687 [I] Mle-----------: Failed to process Advertisement: Drop
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.688 [N] Mle-----------: Attach attempt 0, BetterPartition 
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.688 [I] Mle-----------: AttachState Start -> ParentReq
Dec 18 10:40:53 otbr-agent[783]: 01:27:23.688 [I] Mle-----------: Send Parent Request to routers (ff02:0:0:0:0:0:0:2)
Dec 18 10:40:53 otbr-agent[783]: 01:27:24.438 [I] Mle-----------: Send Parent Request to routers (ff02:0:0:0:0:0:0:2)
Dec 18 10:40:54 otbr-agent[783]: 01:27:24.839 [I] Mle-----------: Receive Parent Response (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.188 [I] Mle-----------: Send Child ID Request (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.189 [I] Mle-----------: AttachState ParentReq -> ChildIdReq
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.285 [I] Mle-----------: Receive Child ID Response (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.286 [N] Mle-----------: Role leader -> detached
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.286 [I] Mle-----------: AttachState ChildIdReq -> Idle
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.286 [N] Mle-----------: RLOC16 6000 -> 3c01
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.293 [N] Mle-----------: Role detached -> child
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.293 [N] Mle-----------: Sending message to inform previous parent 0x7000
Dec 18 10:40:54 otbr-agent[783]: 01:27:25.308 [I] Settings------: ... pid:0x595f5308, mlecntr:0x5e49, maccntr:0x57e3, mliid:1a6d1c20f92217a4}
Dec 18 10:40:55 otbr-agent[783]: 01:27:25.786 [I] Mle-----------: Receive Data Response (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:40:58 otbr-agent[783]: 01:27:29.403 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)

12-18 10:41:33.995  7811 10113 D CTL     : Performing next commissioning step 'SendComplete'
12-18 10:41:34.713  7811 10113 D CTL     : Performing next commissioning step 'Cleanup'

Dec 18 10:41:44 otbr-agent[783]: 01:28:15.480 [I] Mle-----------: Attempt to become router
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.480 [I] Mle-----------: Send Address Solicit (fd87:dcc9:e890:9861:0:ff:fe00:3c00)
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.510 [I] Mle-----------: Receive Address Reply (fd87:dcc9:e890:9861:0:ff:fe00:3c00)
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.511 [N] Mle-----------: RLOC16 3c01 -> 6000
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.517 [N] Mle-----------: Role child -> router
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.517 [N] Mle-----------: Partition ID 0x595f5308
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.517 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.525 [I] Settings------: ... pid:0x595f5308, mlecntr:0x5e4a, maccntr:0x5809, mliid:1a6d1c20f92217a4}
Dec 18 10:41:44 otbr-agent[783]: 01:28:15.529 [I] Mle-----------: Delay Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.700 [I] Mle-----------: Receive Link Request (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.700 [I] Mle-----------: Send Link Accept (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.716 [I] Mle-----------: Receive Data Response (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.716 [I] Mle-----------: Remove Delayed Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.716 [I] Mle-----------: Delay Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:45 otbr-agent[783]: 01:28:15.787 [I] Mle-----------: Send delayed message (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:45 otbr-agent[783]: 01:28:16.229 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:45 otbr-agent[783]: 01:28:16.352 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:41:47 otbr-agent[783]: 01:28:17.842 [I] Mle-----------: Receive Data Response (fe80:0:0:0:3407:28ea:88a8:b90d)
Dec 18 10:41:47 otbr-agent[783]: 01:28:17.843 [I] Mle-----------: Delay Data Response (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:47 otbr-agent[783]: 01:28:18.029 [I] Mle-----------: Send delayed message (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:47 otbr-agent[783]: 01:28:18.173 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:41:47 otbr-agent[783]: 01:28:18.498 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:51 otbr-agent[783]: 01:28:21.842 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:56 otbr-agent[783]: 01:28:26.562 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:41:58 otbr-agent[783]: 01:28:28.882 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:42:10 otbr-agent[783]: 01:28:40.733 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:42:15 otbr-agent[783]: 01:28:45.871 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:42:35 otbr-agent[783]: 01:29:05.970 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:42:42 otbr-agent[783]: 01:29:12.628 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:43:11 otbr-agent[783]: 01:29:41.788 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:43:40 otbr-agent[783]: 01:30:11.234 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
Dec 18 10:43:45 otbr-agent[783]: 01:30:15.617 [I] Mle-----------: Receive Advertisement (fe80:0:0:0:3407:28ea:88a8:b90d,0x3c00)
Dec 18 10:44:21 otbr-agent[783]: 01:30:51.970 [I] Mle-----------: Send Advertisement (ff02:0:0:0:0:0:0:1)
```

**Questions:**
1. Why when the thread network is created, the border router sends "Link Request" I mean which case of the specs is it: "Initial Router Synchronization", "New Router Neighbor Synchronization", "Router Synchronization after Reset". (first or last one ?)
2. In the specs it says, Link Request should be retry MLE_MAX_TRANSMISSION_COUNT (3), but in my traces I see 5 retries, is it normal ? (most of the time, it's only 3)
3. As I understand, "Link request" is to identify the neighbor and available network partitions. In our case there are no partitions because we create a new thread network and it have no devices (it is also confirmed by the fact that all Link Request timeout). But in such cases, as we know that there are no network partitions/neighbor, why did we still try to do the "Network Attaching" process with multiple "Parent request" with retries ? (If there are no neighbor, then I think there are no parent? Or I miss something?)
4. What is the best way to know that the border router is ready in the case of the first network creation ?
6. Shall we use the "state" indicator ? Like "child","router" or "leader" means initialised.
7. Is there another indicator ?

Full logs:
[otbr-agent.txt](https://github.com/openthread/ot-br-posix/files/13705692/otbr-agent.txt)
[matter-commissioner.txt](https://github.com/openthread/ot-br-posix/files/13705697/matter-commissioner.txt)

**Comment by jwhui** [comment, marked as answer]
> 1. Why when the thread network is created, the border router sends "Link Request" I mean which case of the specs is it: "Initial Router Synchronization", "New Router Neighbor Synchronization", "Router Synchronization after Reset". (first or last one ?)

Just configuring a new network does not invalidate the other state that indicates the device was previously in the Leader role. I would suggest using the `factoryreset` CLI command to clear out the other state as well.

> 2. In the specs it says, Link Request should be retry MLE_MAX_TRANSMISSION_COUNT (3), but in my traces I see 5 retries, is it normal ? (most of the time, it's only 3)

Thread 1.3.0 sets `MLE_MAX_CRITICAL_TRANSMISSION_COUNT` to 6.

> 3. As I understand, "Link request" is to identify the neighbor and available network partitions. In our case there are no partitions because we create a new thread network and it have no devices (it is also confirmed by the fact that all Link Request timeout). But in such cases, as we know that there are no network partitions/neighbor, why did we still try to do the "Network Attaching" process with multiple "Parent request" with retries ? (If there are no neighbor, then I think there are no parent? Or I miss something?)

Try using the `factoryreset` CLI command.

> 4. What is the best way to know that the border router is ready in the case of the first network creation ?
> 5. Shall we use the "state" indicator ? Like "child","router" or "leader" means initialised.
> 6. Is there another indicator ?

Observing the Thread role state is a good indicator if when the device is attached to a Thread network.

**Comment by caipiblack** [comment]
Thanks you so much! 

With a factoryreset we don't have all the steps with 6x Link Request in timeout of 5s (+/-10%) so it reduce drasticly the network creation time.

```
ot-ctl factoryreset
sleep 3
ot-ctl thread stop
ot-ctl dataset init new
ot-ctl dataset channel $1
ot-ctl dataset commit active
ot-ctl ifconfig up
ot-ctl thread start
wait_ot_ctl_state
```

`wait_ot_ctl_state` is a function that execute ot-ctl state in loop every one seconds and wait for "child","leader" or "router" result (with timeout)

So now when we create the thread networks, it takes 10.8s to be ready:
- We loose 3 seconds during the factoryreset
- We loose 5.25s (0.75x2 + 1.25x3) seconds for the parent req in timeouts (+1s for detection)
- The rest is loss somewhere.

**Comment by caipiblack** [comment]
Hello,

I have a new question, I'm going to ask here because it is highly related.

I try to understand why the thread device is doing the DNS Update (addressQuery + dnsUpdate) only 7seconds after being attached.

Do you have any ideas ?

Here there is a sequence diagram based on otbr-agent logs:

![appairage_eve](https://github.com/openthread/openthread/assets/29680146/d2699f2c-da12-47e9-961c-51e65f0fe244)

Attached you will find the full logs of otbr-agent and the sources of this plantuml diagram.

[appairage_eve.txt](https://github.com/openthread/openthread/files/13779907/appairage_eve.txt)
[otbr-agent-logs.txt](https://github.com/openthread/openthread/files/13779910/otbr-agent-logs.txt)

Also, on the past we had the logs of the ot-ctl commands in the otbr-agent logs, but now (since long time), we don't have it.

Did you know how to display the commands executed by ot-ctl in otbr-agent logs ?

Note:

The logs that you see like this `execute_ot_ctl_2: ot-ctl state...` comes from a script that we made to execute ot-ctl commands, it also put logs in otbr-agent logs by calling something like `logger -t otbr-agent "some text"` but in the past it was possible to do it "without this trick"

**Comment by jwhui** [comment]
> I try to understand why the thread device is doing the DNS Update (addressQuery + dnsUpdate) only 7seconds after being attached.
> 
> Do you have any ideas ?

Take a look at https://github.com/openthread/openthread/pull/8318, which may help avoid the need to resolve the SRP Server's address.

> Also, on the past we had the logs of the ot-ctl commands in the otbr-agent logs, but now (since long time), we don't have it.
> 
> Did you know how to display the commands executed by ot-ctl in otbr-agent logs ?

I believe this is due to https://github.com/openthread/openthread/pull/8994

**Comment by caipiblack** [comment]
Thanks! With your link I also find related issues/discussions:

- https://github.com/openthread/openthread/issues/9416
- https://github.com/orgs/openthread/discussions/8202

The problem explained here: https://github.com/openthread/openthread/issues/9416#issue-1890834715 is exactly what I see, but I'm not sure if the source of the problem is the same, because in my case I don't have the `AddrResolver--: Timed out waiting for address notification` in otbr-agent logs. 

But it's not clear to me if their logs are "device side" or "otbr-agent side", it's possible that it's "device side".

Also it's not clear to me if this PR must be present "device side" or "otbr-agent side": https://github.com/openthread/openthread/pull/8318 

if it's otbr-agent side, I will try it.

**Reply by jwhui** [reply]
The enhancement in https://github.com/openthread/openthread/pull/8318 avoid the need to send an Address Query message in certain situations. In your scenario, the device side is sending the Address Query, so this optimization needs to be on the device side.
